### PR TITLE
feat(evm): EIP-7002 withdrawal requests

### PR DESCRIPTION
**Pull request by shekhirin**
Implements the [EIP-7002](https://eips.ethereum.org/EIPS/eip-7002) post-block system contract call. Tests are pending, will be added in the next PR.

### CODE REVIEW

File: crates/ethereum/evm/src/execute.rs [hunk 1]
```diff
@@ -258,23 +266,23 @@ where
...
+    /// Returns the receipts of the transactions in the block, the total gas used and the list of
+    /// EIP-7685 [requests](Request).
     ///
     /// Returns an error if execution fails or receipt verification fails.
     fn execute_and_verify(
         &mut self,
         block: &BlockWithSenders,
         total_difficulty: U256,
-    ) -> Result<(Vec<Receipt>, u64), BlockExecutionError> {
+    ) -> Result<(Vec<Receipt>, Vec<Request>, u64), BlockExecutionError> {
```
Comment (mattsse) [inline crates/ethereum/evm/src/execute.rs:277]: can we add a type for this

File: crates/evm/src/execute.rs [hunk 2]
```diff
@@ -80,6 +81,8 @@ pub struct BlockExecutionOutput<T> {
     pub state: BundleState,
     /// All the receipts of the transactions in the block.
     pub receipts: Vec<T>,
+    /// All the EIP-7685 requests of the transactions in the block.
+    pub requests: Vec<Request>,
```
Comment (mattsse) [inline crates/evm/src/execute.rs:85]: this is fine for now

File: crates/ethereum/evm/src/execute.rs [hunk 3]
```diff
@@ -118,16 +122,18 @@ where
...
-    /// It does __not__ apply post-execution changes.
-    fn execute_pre_and_transactions<Ext, DB>(
+    /// It does __not__ apply post-execution changes that do not require an [EVM](Evm), for that see
+    /// [EthBlockExecutor::post_execution].
+    fn execute_state_transitions<Ext, DB>(
         &self,
         block: &BlockWithSenders,
-        mut evm: Evm<'_, Ext, &mut State<DB>>,
-    ) -> Result<(Vec<Receipt>, u64), BlockExecutionError>
+        evm: &mut Evm<'_, Ext, &mut State<DB>>,
```
Comment (gakonst) [inline crates/ethereum/evm/src/execute.rs:135]: @mattsse is this OK to borrow mutably here

File: crates/ethereum/evm/src/execute.rs [hunk 4]
```diff
@@ -197,7 +202,10 @@ where
             .into())
         }
 
-        Ok((receipts, cumulative_gas_used))
+        let withdrawal_requests =
+            post_block_withdrawal_requests(&self.chain_spec, block.timestamp, evm)?;
+
+        Ok((receipts, cumulative_gas_used, withdrawal_requests))
```
Comment (gakonst) [inline crates/ethereum/evm/src/execute.rs:208]: TODO: we should convert this to a struct later

File: crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs [hunk 5]
```diff
@@ -38,16 +38,18 @@ pub struct BundleStateWithReceipts {
...
 impl From<BatchBlockExecutionOutput> for BundleStateWithReceipts {
     fn from(value: BatchBlockExecutionOutput) -> Self {
-        let BatchBlockExecutionOutput { bundle, receipts, first_block } = value;
+        let BatchBlockExecutionOutput { bundle, receipts, requests: _, first_block } = value;
         Self { bundle, receipts, first_block }
     }
 }
 
 // TODO(mattsse): unify the types, currently there's a cyclic dependency between
+#[cfg(any(test, feature = "test-utils"))]
```
Comment (mattsse) [inline crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs:47]: we still need this I believe

File: crates/ethereum/evm/src/execute.rs (hunk 1 above)
Comment (gakonst) [inline crates/ethereum/evm/src/execute.rs:277]: nit: ordering, it's Vec<Request> here 2nd arg whereas on the function above it's 3rd arg

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/primitives/src/revm/env.rs [hunk 6]
```diff
@@ -132,24 +132,59 @@ pub fn tx_env_with_recovered(transaction: &TransactionSignedEcRecovered) -> TxEn
...
+/// validations), call the contract as `SYSTEM_ADDRESS`.
+pub fn fill_tx_env_with_withdrawal_requests_contract_call(env: &mut Env) {
+    fill_tx_env_with_system_contract_call(
+        env,
+        alloy_eips::eip7002::SYSTEM_ADDRESS,
+        WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS,
+        Bytes::new(),
+    );
+}
+
```
Comment (gakonst) [inline crates/primitives/src/revm/env.rs:164]: great stuff

File: crates/revm/src/state_change.rs [hunk 7]
```diff
@@ -158,3 +168,65 @@ pub fn insert_post_block_withdrawals_balance_increments(
         }
     }
 }
+
+/// Applies the post-block call to the EIP-7002 withdrawal requests contract.
+///
+/// If Prague is not active at the given timestamp, then this is a no-op, and an empty vector is
+/// returned. Otherwise, the withdrawal requests are returned.
+#[inline]
+pub fn post_block_withdrawal_requests<EXT, DB: Database + DatabaseCommit>(
```
Comment (gakonst) [inline crates/revm/src/state_change.rs:177]: needs unit test but OK with moving fast

**Review by gakonst — COMMENTED** [review]
good start

File: crates/evm/src/execute.rs (hunk 2 above)
Comment (shekhirin) [reply crates/evm/src/execute.rs:85]: how would you improve it?

File: crates/ethereum/evm/src/execute.rs (hunk 1 above)
Comment (shekhirin) [reply crates/ethereum/evm/src/execute.rs]: added

File: crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs (hunk 5 above)
Comment (shekhirin) [reply crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs:47]: yeah but it doesn't include the requests in the bundle state now, so for now I gated it behind the cfg and added a todo

File: crates/ethereum/evm/src/execute.rs (hunk 4 above)
Comment (shekhirin) [reply crates/ethereum/evm/src/execute.rs]: converted now

File: crates/ethereum/evm/src/execute.rs (hunk 1 above)
Comment (shekhirin) [reply crates/ethereum/evm/src/execute.rs]: fixed

File: crates/revm/src/state_change.rs (hunk 7 above)
Comment (shekhirin) [reply crates/revm/src/state_change.rs:177]: gonna add it after we merge this PR

File: crates/ethereum/evm/src/execute.rs (hunk 3 above)
Comment (shekhirin) [reply crates/ethereum/evm/src/execute.rs]: no need

File: crates/ethereum/evm/src/execute.rs [hunk 8]
```diff
@@ -105,6 +107,14 @@ where
     }
 }
 
+/// Helper type for the output of executing a block.
+#[derive(Debug, Clone)]
+struct EthExecuteOutput {
```
Comment (mattsse) [inline crates/ethereum/evm/src/execute.rs:112]: I will combine this with the other output types separately

**Review by mattsse — APPROVED** [review]

**Pull request merged**